src/Adapter/Discount/Repository/DiscountRepository.php

```diff
@@ -50,6 +50,7 @@ public function __construct(
         protected readonly Connection $connection,
         protected readonly string $dbPrefix
     ) {
+        $this->cartRuleValidator->setDiscountRepository($this);
```

Comment:
I'm not sure if it would solve the cirular dependency problem, but the Syfony DI allows to inject services via setter methods, maybe it's worth trying This way the DI responsibility lays on the framework and not this repository service If not I understand the need and it's acceptable this way

Reply:
https://symfony.com/doc/6.4/service_container/calls.html